### PR TITLE
Fix crasher when checking encoding of pseudo files during testing

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
Fix crasher when checking file encoding of pseudo files during testing

### Linked issues
None

### Functionality 
None

### Tests
- [x] manually tested

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/notebooks/sources.py [hunk 1]
```diff
@@ -165,10 +166,22 @@ def __init__(self, ctx: LinterContext, path: Path, content: str | None = None):
...
-        return self._path.read_text(encoding) if self._content is None else self._content
+        if self._content is None:
+            self._content = self._path.read_text(self._guess_encoding())
+        return self._content
+
+    def _guess_encoding(self):
+        path = self._path.as_posix()
+        count = min(32, os.path.getsize(path))
+        with open(path, 'rb') as _file:
+            raw = _file.read(count)
```
Comment (asnare) [inline src/databricks/labs/ucx/source_code/notebooks/sources.py:179]:
```suggestion
        with self._path.open('rb') as _file:
            raw = _file.read(4)
```
(I think all BOMs are 4 bytes or less.)

File: src/databricks/labs/ucx/source_code/notebooks/sources.py [hunk 2]
```diff
@@ -165,10 +166,22 @@ def __init__(self, ctx: LinterContext, path: Path, content: str | None = None):
...
+            if raw.startswith(codecs.BOM_UTF8):
+                return 'utf-8-sig'
+            if raw.startswith(codecs.BOM_UTF16_LE) or raw.startswith(codecs.BOM_UTF16_BE):
+                return 'utf-16'
```
Comment (asnare) [inline src/databricks/labs/ucx/source_code/notebooks/sources.py:183]: Here I think we should either cover all the BOMs or add a comment explaining why we only support this particular subset of BOM-marked files.

File: tests/unit/source_code/notebooks/test_sources.py [hunk 3]
```diff
@@ -21,6 +22,49 @@ def test_file_linter_lints_not_yet_supported_language(path, migration_index):
...
+    assert linter.source_code() is not None
+
+
+def test_checks_encoding_of_file_with_utf16_le_bom(migration_index):
+    with (tempfile.NamedTemporaryFile() as tf):
+        data = bytearray()
+        data.append(0xFF)
+        data.append(0xFE)
+        for b in "a = 12".encode('utf-16-le'):
+            data.append(b)
```
Comment (asnare) [inline tests/unit/source_code/notebooks/test_sources.py:48]: This is fine, although the alternative often seen is the `b'\x__' syntax, which would mean something like:
```python
data = b"\xFF\xFE" + "a = 12".encode("utf-16-le")
```

File: tests/unit/source_code/notebooks/test_sources.py [hunk 4]
```diff
@@ -21,6 +22,49 @@ def test_file_linter_lints_not_yet_supported_language(path, migration_index):
...
+    assert linter.source_code() == "a=b"
+
+
+def test_checks_encoding_of_file_with_utf8_bom(migration_index, mock_path_lookup):
+    path = mock_path_lookup.resolve(Path("file_with_bom.py"))
+    linter = FriendFileLinter(LinterContext(migration_index), path)
+    assert linter.source_code() is not None
+
+
+def test_checks_encoding_of_file_with_utf16_le_bom(migration_index):
```
Comment (asnare) [inline tests/unit/source_code/notebooks/test_sources.py:42]: I think this and the next test can be trivially parameterised.

File: src/databricks/labs/ucx/source_code/notebooks/sources.py [hunk 5]
```diff
@@ -165,10 +166,22 @@ def __init__(self, ctx: LinterContext, path: Path, content: str | None = None):
...
-    @cached_property
+    @property
     def _source_code(self) -> str:
-        encoding = locale.getpreferredencoding(False)
-        return self._path.read_text(encoding) if self._content is None else self._content
+        if self._content is None:
+            self._content = self._path.read_text(self._guess_encoding())
+        return self._content
+
+    def _guess_encoding(self):
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebooks/sources.py:175]: done

File: src/databricks/labs/ucx/source_code/notebooks/sources.py (hunk 2 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebooks/sources.py:183]: done

File: tests/unit/source_code/notebooks/test_sources.py (hunk 3 above)
Comment (ericvergnaud) [reply tests/unit/source_code/notebooks/test_sources.py:48]: done, thanks

File: tests/unit/source_code/notebooks/test_sources.py (hunk 4 above)
Comment (ericvergnaud) [reply tests/unit/source_code/notebooks/test_sources.py:42]: done

**Review by asnare — APPROVED** [review]

File: src/databricks/labs/ucx/source_code/notebooks/sources.py [hunk 6]
```diff
@@ -165,10 +165,25 @@ def __init__(self, ctx: LinterContext, path: Path, content: str | None = None):
...
+        # some files encode a unicode BOM (byte-order-mark), so let's use that if available
+        path = self._path.as_posix()
+        with open(path, 'rb') as _file:
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/notebooks/sources.py:177]:
```suggestion
        with self._path.open('rb') as _file:
```

File: tests/unit/source_code/notebooks/test_sources.py [hunk 7]
```diff
@@ -21,6 +23,36 @@ def test_file_linter_lints_not_yet_supported_language(path, migration_index):
     assert [advice.code for advice in advices] == ["unsupported-language"]
 
 
+class FriendFileLinter(FileLinter):
```
Comment (JCZuurmond) [inline tests/unit/source_code/notebooks/test_sources.py:26]: Why `Friend`?

File: tests/unit/source_code/notebooks/test_sources.py [hunk 8]
```diff
@@ -21,6 +23,36 @@ def test_file_linter_lints_not_yet_supported_language(path, migration_index):
...
+        (codecs.BOM_UTF32_BE, "utf-32-be"),
+    ],
+)
+def test_checks_encoding_of_file_with_bom(migration_index, bom, encoding):
+    with tempfile.NamedTemporaryFile() as temp:
+        data = bom + "a = 12".encode(encoding)
+        temp.write(data)
+        temp.flush()
+        linter = FriendFileLinter(LinterContext(migration_index), Path(temp.name))
+        assert linter.source_code() == "a = 12"
```
Comment (JCZuurmond) [inline tests/unit/source_code/notebooks/test_sources.py:53]: The `tmp_path` pytest fixture is commonly used for temporary files in tests

```suggestion
def test_checks_encoding_of_file_with_bom(migration_index, bom, encoding, tmp_path):
    path = tmp_path / "file.py"
    path.write_bytes(bom + "a = 12".encode(encoding))
    linter = FriendFileLinter(LinterContext(migration_index), path)
    assert linter.source_code() == "a = 12"
```

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
Some nits

File: tests/unit/source_code/notebooks/test_sources.py (hunk 7 above)
Comment (asnare) [reply tests/unit/source_code/notebooks/test_sources.py:26]: We have linting rules that block access to protected members. It's normal (and conventional) in most languages for tests to have better-than-public access to the internals of the thing being tested: unit tests are often (intentionally) tightly coupled to the thing under test and are necessarily aware of the implementation and internals.

In C++ there's a concept of "friend" classes: these are allowed to access protected and private attributes. (The package-private concept of Java is similar, and a major reason that unit tests are conventionally placed in the same package as the classes being tested.)

So tying things together: `Friend` is a hint that it's purpose is to access protected attributes and allow them to be used during testing.

Comment (JCZuurmond) [reply tests/unit/source_code/notebooks/test_sources.py:26]: Thanks for the explanation! Learned something new here :)

File: src/databricks/labs/ucx/source_code/notebooks/sources.py (hunk 6 above)
Comment (nfx) [reply src/databricks/labs/ucx/source_code/notebooks/sources.py:177]: +1

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/source_code/notebooks/sources.py [hunk 9]
```diff
@@ -165,10 +165,25 @@ def __init__(self, ctx: LinterContext, path: Path, content: str | None = None):
         self._path: Path = path
         self._content = content
 
-    @cached_property
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebooks/sources.py:168]: Done (it's already cached in self._content hence the memory optimization..)

File: tests/unit/source_code/notebooks/test_sources.py (hunk 8 above)
Comment (ericvergnaud) [reply tests/unit/source_code/notebooks/test_sources.py:53]: done

File: src/databricks/labs/ucx/source_code/notebooks/sources.py (hunk 6 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/notebooks/sources.py:178]: done

**Review by nfx — APPROVED** [review]
lgtm

**Comment by ericvergnaud** [comment]
> @ericvergnaud please fix `make fmt`

@nfx done

**Pull request merged**